File: test/unit/cpp_integration/function_return.ts
import { ValueReturner } from "./declarations/function_return";

const returner = new ValueReturner();

const smallPod16 = returner.getSmallPod16();
console.assert(smallPod16.getValue() === 1, "Small pod 16 test failed");

const smallPod32 = returner.getSmallPod32();
console.assert(smallPod32.getValue() === 2, "Small pod 32 test failed");

const smallPodWithVirtualDestructor = returner.getSmallWithVirtualDestructor();
console.assert(smallPodWithVirtualDestructor.getValue() === 3, "Small pod with virtual destructor test failed");

/* @todo:
    "If the size of an object is larger than four eightbytes, or it contains unaligned fields, it has class MEMORY"

const smallUnaligned = returner.getSmallUnaligned();
console.assert(smallUnaligned.getValue() === 4, "Small unaligned test failed");
*/

const large = returner.getLarge();
console.assert(large.getValue() === 5, "Large test failed");

File: test/src/cpp_integration/cpp/function_return.cpp
#include "function_return.h"

SmallPod16::SmallPod16() {}
int16_t SmallPod16::getValue() const { return _; }

SmallPod32::SmallPod32() {}
int32_t SmallPod32::getValue() const { return _; }

SmallWithVirtualDestructor::SmallWithVirtualDestructor() {}
SmallWithVirtualDestructor::~SmallWithVirtualDestructor() {}

int16_t SmallWithVirtualDestructor::getValue() const { return _; }

SmallUnaligned::SmallUnaligned() {}
int16_t SmallUnaligned::getValue() const { return _1; }

Large::Large() {}
int64_t Large::getValue() const { return _1; }

ValueReturner::ValueReturner() {}

SmallPod16 ValueReturner::getSmallPod16() const { return {}; }
SmallPod32 ValueReturner::getSmallPod32() const { return {}; }

SmallWithVirtualDestructor
ValueReturner::getSmallWithVirtualDestructor() const {
  return {};
}
SmallUnaligned ValueReturner::getSmallUnaligned() const { return {}; }
Large ValueReturner::getLarge() const { return {}; }
